File: magmac/src/c/com/meti/PathSource.c
#include "PathSource.c"public record PathSource(Path root, Path path) {
    public List<String> findPackage() {
        var list = new ArrayList<String>();
        var parent = root.toAbsolutePath().relativize(path.toAbsolutePath()).getParent();
        if(parent == null) return Collections.emptyList();

        for (int i = 0; i < parent.getNameCount(); i++) {
            list.add(parent.getName(i).toString());
        }
        return list;
    }

    public String findName() {
        var name = path.getName(path.getNameCount() - 1).toString();
        var separator = name.indexOf('.');
        return name.substring(0, separator);
    }

    public String read() throws IOException {
        return Files.readString(path);
    }
}

File: magmac/src/java/com/meti/PathSource.java
package com.meti;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

public record PathSource(Path root, Path child) {
    String findName() {
        var fileName = child.getFileName().toString();
        var separator = fileName.indexOf('.');
        return fileName.substring(0, separator);
    }

    List<String> findNamespace() {
        var list = new ArrayList<String>();
        var relative = root.relativize(child);
        for (int i = 0; i < relative.getNameCount() - 1; i++) {
            list.add(relative.getName(i).toString());
        }
        return list;
    }

    public String read() throws IOException {
        return Files.readString(child);
    }
}
